presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroup.java

```diff
@@ -0,0 +1,486 @@
+                group.parent.get().dirtySubGroups.add(group);
+                group = group.parent.get();
+            }
+            updateEligiblility();
+            executor.execute(query::start);
+        }
+    }
+
+    private void queryFinished(QueryExecution query)
+    {
+        synchronized (root) {
+            if (!runningQueries.contains(query) && !queuedQueries.contains(query)) {
+                // Query has already been cleaned up
+                return;
+            }
+            if (runningQueries.contains(query)) {
+                runningQueries.remove(query);
+                ResourceGroup group = this;
+                while (group.parent.isPresent()) {
+                    group.parent.get().descendantRunningQueries--;
+                    group = group.parent.get();
+                }
+            }
+            else {
+                queuedQueries.remove(query);
+                ResourceGroup group = this;
+                while (group.parent.isPresent()) {
+                    group.parent.get().descendantQueuedQueries--;
+                    group = group.parent.get();
+                }
+            }
+            updateEligiblility();
+        }
+    }
+
+    protected void internalRefreshStats()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to refresh stats");
+        synchronized (root) {
+            if (subGroups.isEmpty()) {
+                cachedMemoryUsageBytes = 0;
+                for (QueryExecution query : runningQueries) {
+                    cachedMemoryUsageBytes += query.getTotalMemoryReservation();
+                }
+            }
+            else {
+                for (Iterator<ResourceGroup> iterator = dirtySubGroups.iterator(); iterator.hasNext(); ) {
+                    ResourceGroup subGroup = iterator.next();
+                    cachedMemoryUsageBytes -= subGroup.cachedMemoryUsageBytes;
+                    subGroup.internalRefreshStats();
+                    cachedMemoryUsageBytes += subGroup.cachedMemoryUsageBytes;
+                    if (!subGroup.isDirty()) {
+                        iterator.remove();
+                    }
+                }
+            }
+        }
+    }
+
+    protected boolean internalStartNext()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to find next query");
+        synchronized (root) {
+            if (!canRunMore()) {
+                return false;
+            }
+            QueryExecution query = queuedQueries.poll();
+            if (query != null) {
+                startInBackground(query);
+                return true;
+            }
+
+            // Remove even if the sub group still has queued queries, so that it goes to the back of the queue
+            ResourceGroup subGroup = eligibleSubGroups.poll();
+            if (subGroup == null) {
+                return false;
+            }
+            checkState(subGroup.internalStartNext(), "Eligible sub group had no queries to run");
```

Comment:
This would be easier to read if you assign `subGroup.internalStartNext()` to a variable and then check separately.  On my first read I did not notice the side effects here